dockerworker/artifact.go

```diff
@@ -0,0 +1,151 @@
+// +build docker
+
+package dockerworker
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/mattetti/filebuffer"
+	"github.com/mitchellh/ioprogress"
+	"github.com/taskcluster/taskcluster-client-go/tcqueue"
+	artifact "github.com/taskcluster/taskcluster-lib-artifact-go"
+)
+
+// ScheduleReclaim run a goroutine that will perform reclaims in the background.
+func (d *DockerWorker) ScheduleReclaim(claim *tcqueue.TaskClaimResponse) {
+	takenUntil := claim.TakenUntil
+
+	go func() {
+		for {
+			select {
+			case <-time.After(time.Time(takenUntil).Sub(time.Now())):
+				reclaim, err := d.queue.ReclaimTask(claim.Status.TaskID, fmt.Sprint(claim.RunID))
+				if err != nil {
+					d.TaskLogger.Printf("Exiting reclaim loop: %v", err)
+					return
+				}
+				takenUntil = reclaim.TakenUntil
+			case <-d.Context.Done():
+				return
+			}
+		}
+	}()
+}
```

Comment:
Why is this logic in the docker package? Reclaim logic already exists in the worker, and doesn't need to be specialised for the docker engine.

Reply:
The duplicated functions are used only to test the package (you could argue that they shouldn't be exported then, but it causes no harm either), as they simpler versions and more generic version of the functionality provided by the main package. The integrated code keeps using the generic-worker provided functions.

Reply:
The test framework already provides mechanisms to submit tasks, upload artifacts, claim tasks etc, so there is no reason to duplicate this functionality. Please have a look at existing tests to see how this is done. These functions are independent of the engine being used, and the existing code does not need any alteration to be used.

Reply:
IIUC, the test frame aims black box tests, these utility functions are for white box tests. Since these are used only for tests, I don't think it is a big concern such a small duplication in name of code organization, as long as it doesn't cause regressions. We can always refactor it later.